src/Raven.Server/Documents/Handlers/TimeSeriesHandler.cs

```diff
@@ -161,22 +172,46 @@ protected override int ExecuteCmd(DocumentsOperationContext context)
                     return 0;
 
                 var tss = _database.DocumentsStorage.TimeSeriesStorage;
+                var holder = new TimeSeriesStorage.Reader.SingleResult();
 
-                if (_batch.Appends != null)
+                if (_appendDictionary != null)
                 {
-                    foreach (var append in _batch.Appends)
+                    foreach (var kvp in _appendDictionary)
                     {
                         LastChangeVector = tss.AppendTimestamp(context,
                             _batch.Id,
                             docCollection,
-                            append.Name,
-                            append.Timestamp,
-                            new Span<double>(append.Values),
-                            append.Tag,
-                            fromReplication: false
+                            kvp.Key,
+                            kvp.Value.Values.Select(x =>
+                            {
+                                holder.Values = x.Values;
+                                holder.Tag = context.GetLazyString(x.Tag);
+                                holder.TimeStamp = x.Timestamp;
+                                return holder;
+                            })
                         );
 
-                        changes++;
+                        changes += kvp.Value.Count;
+                    }
+
+                    _appendDictionary.Clear();
+                }
+                else if (_singleValue != null)
+                {
+                    LastChangeVector = tss.AppendTimestamp(context,
+                        _batch.Id,
+                        docCollection,
+                        _singleValue.Name,
+                        Single(_singleValue));
+
+                    changes++;
+
+                    IEnumerable<TimeSeriesStorage.Reader.SingleResult> Single(AppendTimeSeriesOperation element)
```

Comment:
I don't really like this approach, can we accept `AppendTimeSeriesOperation` directly and convert internally?